.clang-format

```diff
@@ -0,0 +1,283 @@
+
+# Use the same indentation level as for the switch statement.
+# Switch statement body is always indented one level more than case labels;
+# this is followin K&R C-style (case labels are really just goto labels).
+#
+# switch (foo) {
+# case 1:
+# 	bar();
+# 	break;
+# default:
+# 	baz();
+# }
+#
+IndentCaseLabels: false
+
+# Don't insert a space after a cast
+#
+# x = (int32)y;    NOT    x = (int32) y;
+#
+SpaceAfterCStyleCast: false
+
+# Insert spaces before and after assignment operators
+#
+# int a = 5;    NOT    int a=5;
+# a += 42;             a+=42;
+#
+SpaceBeforeAssignmentOperators: true
+
+# Put a space before opening parentheses only after control statement keywords.
+#
+# void f()
+# {
+#         if (true) {
+#                 f();
+#         }
+# }
+#
+SpaceBeforeParens: ControlStatements
+
+# Don't insert spaces inside empty '()'
+#
+SpaceInEmptyParentheses: false
+
+# The number of spaces before trailing line comments (// - comments).
+# This does not affect trailing block comments (/* - comments).
+#
+SpacesBeforeTrailingComments: 1
+
+# Don't insert spaces in casts
+#
+# x = (int32) y;  NOT   x = ( int32 ) y;
+#
+SpacesInCStyleCastParentheses: false
+
+# Don't insert spaces after '(' or before ')'
+#
+# f(arg);         NOT    f( arg );
+#
+SpacesInParentheses: false
+
+# Don't insert spaces after '[' or before ']'
+#
+# int a[5];       NOT    int a[ 5 ];
+#
+SpacesInSquareBrackets: false
+
+# Insert a space after '{' and before '}' in struct initializers
+#
+# This is native C++11 style, but it looks very weird.
+# TODO Investigate if it has any tangible benefits.
+#
+# Cpp11BracedListStyle: false
+
+# A list of macros that should be interpreted as foreach loops instead of as
+# function calls.
+#
+# TODO Currently unused, but left as an example in case we'll need such macros.
+#
+# ForEachMacros:
+#   - 'for_each_abbrev'
+#   - 'list_for_each_dir'
+
+# The maximum number of consecutive empty lines to keep.
+#
+MaxEmptyLinesToKeep: 1
+
+# No empty line at the start of a block.
+#
+KeepEmptyLinesAtTheStartOfBlocks: false
+
+# Line breaking penalties
+#
+# This decides what order things should be done if a line is too long
+#
+# clang-format iterates through various versions the long line can be formatted 
+# and selects the one with smallest penalty score.
+#
+# Small ExcessCharacter penalty prevents breaking line is longer than
+# column limit by only few characters.
+#
+PenaltyBreakAssignment: 100
+PenaltyBreakBeforeFirstCallParameter: 100
+PenaltyBreakComment: 10
+PenaltyBreakFirstLessLess: 0
+PenaltyBreakString: 110
+PenaltyExcessCharacter: 3
+PenaltyReturnTypeOnItsOwnLine: 200
+
+# Don't sort "#include" declarations.
+#
+# clang-format has very rich customization options for grouping and sorting
+# "include" directives, but there are still edge cases, so we still need to
+# rely on developers doing this manually.
+#
+# See https://github.com/dreamer/dosbox-staging/issues/196 for details.
+#
```

Comment:
Would be nice if we could have some logic here:
 - if <ctypes> and <algorithm> present, then don't sort.
 - otherwise re-arrange per X, Y, Z rules.

Likewise, re-arranging the overall blocks of headers (system headers before local headers) would be nice.

Sample with re-arranging just the local headers in `"` quotes might be doable, if there's a distinctions. 

All just nice-to-have, if it can do it.

Reply:
Formatter needs to arrive at includes order deterministically, so include grouping rules depend on sorting :( There's an option to group and sort each group separately and define multiple groups though.

I have a few ideas on how to resolve it so we could enable it in the future, but it will need some experimentation and a separate PR.